src/algebra/group/freiman.lean

```diff
@@ -0,0 +1,405 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.fun_like
+import data.multiset.basic
+
+/-!
+# Freiman homomorphisms
+
+In this file, we define Freiman homomorphisms. A `n`-Freiman homomorphism is a function `f : α → β`
+such that `f (x₁) * ... * f (xₙ) = f (y₁) * ... * f (yₙ)` for all `x₁, ..., xₙ, y₁, ..., yₙ` such
+that `x₁ * ... * xₙ = y₁ * ... * yₙ`. In particular, any `mul_hom` is a Freiman homomorphism.
+
+They are of interest in additive combinatorics.
+
+## Main declaration
+
+* `freiman_hom`: Freiman homomorphism.
+* `add_freiman_hom`: Additive Freiman homomorphism.
+
+## Notation
+
+* `A →*[n] β`: Multiplicative `n`-Freiman homomorphism
+* `α →+[n] β`: Additive`n`-Freiman homomorphism
+
+## References
+
+[Yufei Zhao, *Graph Theory and Additivite Combinatorics][https://yufeizhao.com/gtac/]
```

Comment:
```suggestion
[Yufei Zhao, *18.225: Graph Theory and Additivite Combinatorics][https://yufeizhao.com/gtac/]
```
since that would flag to MIT alum like me that "hey, that's an MIT course!". (and it's part of the page title anyway)